### PR TITLE
Check if NSURLProtocolClient has private callback before calling it

### LINKED ISSUE #2764: NSURLProtocol calls private extension method on NSURLProtocolClient even if it doesn't exist
Using the 0.2.170511-dev-20170531235959 nuget packages we’re seeing the following crash in our app:

```
 	KernelBase.dll!_RaiseException@16()	Unknown
 	vcruntime140_app.dll!__CxxThrowException@8()	Unknown
 	libobjc2.dll!objc_exception_throw(void * object) Line 134	C++
>	Foundation.dll!77835533()	Unknown
 	[Frames below may be incorrect and/or missing, no symbols loaded for Foundation.dll]	
 	ntdll.dll!_RtlpHpLfhSubsegmentCommitBlock@12()	Unknown
 	ntdll.dll!_NtClose@4()	Unknown
 	ntdll.dll!_NtClose@4()	Unknown
 	ntdll.dll!_NtClose@4()	Unknown
```

### PR DESCRIPTION
**Pull request by ms-jihua**
Fixes #2764

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/winobjc/2765)

### CODE REVIEW

File: Frameworks/Foundation/NSURLProtocol_WinHTTP.mm [hunk 1]
```diff
@@ -288,7 +288,8 @@ - (void)startLoading {
             // TODO #2721: If this properly supports streaming bodies instead of flattening the data up front,
             // more accurate callbacks can be done
             NSInteger contentSize = self.request.HTTPBody ? self.request.HTTPBody.length : _flattenedBodyStream.size();
-            if (contentSize > 0) {
+            if (([self.client respondsToSelector:@selector(URLProtocol:didWriteData:totalBytesWritten:totalBytesExpectedToWrite:)]) &&
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSURLProtocol_WinHTTP.mm]: In looking for this issue, I realized that no header declares the protocol method `URLProtocol:didWriteData:...`; could we add one in NSURLProtocolInternal? Ideally, NSURLConnection would conform to this private protocol, and the compiler would warn us if it didn't have the right methods on it. NSURLSessionTask _should_ eventually do this, but it doesn't.

Comment (ms-jihua) [inline Frameworks/Foundation/NSURLProtocol_WinHTTP.mm]: This is literally declared in NSURLProtocolInternal.h...

Comment (rajsesh-msft) [inline Frameworks/Foundation/NSURLProtocol_WinHTTP.mm]: there are other "private" invocations in this file, are these expected to succeed?

Comment (ms-jihua) [inline Frameworks/Foundation/NSURLProtocol_WinHTTP.mm]: What do you mean by other "private" invocations?

**Comment by DHowett-MSFT** [comment]
@ms-jihua this is the entirety of NSURLProtocolInternal.h for me:

```
objc
//******************************************************************************
//
// Copyright (c) Microsoft. All rights reserved.
//
// This code is licensed under the MIT License (MIT).
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//
//******************************************************************************

#pragma once

#import "Starboard/SmartTypes.h"

@interface NSURLProtocol ()
+ (id)_URLProtocolClassForRequest:(id)request;

@end
```

**Comment by ms-jihua** [comment]
Yup, we lack POST/UploadTask tests for NSURLSession - should I add some?

**Comment by ms-jihua** [comment]
@DHowett-MSFT ...I somehow missed the existence of another NSURLProtocolInternal.h in Frameworks/include. I'll coalesce it with the one in Frameworks\Foundation.

**Comment by rajsesh-msft** [comment]
@ms-jihua yes, It would be helpful to have at least one test for POST using NSURLSession.

**Comment by ms-jihua** [comment]
@DHowett-MSFT @rajsesh-msft wait a second, [NSURLSession uploadTaskWith...] functions are still stubs. How did we hit this bug?

**Comment by DHowett-MSFT** [comment]
@ms-jihua two things: 1) users can implement NSURLProtocolClient on their own, and they don't know about this. 2) A data task can still be a post with a body stream if it's created from an NSURLRequest with a body/bodystream. Also, GETs can have bodies just as well as POSTs can. :smile:

**Pull request merged**